### DISCUSSION TITLE
Benchmark - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#2680] [Docs]
# Benchmark - Ultralytics YOLOv8 Docs

Benchmark mode compares speed and accuracy of various YOLOv8 export formats like ONNX or OpenVINO. Optimize formats for speed or accuracy.

https://docs.ultralytics.com/modes/benchmark/

**Comment by MohitDAngrish** [comment]
how do we provide our custom dataset for benchmarking, I don't see an option for providing the dataset info.

**Reply by pderrenger** [reply]
@MohitDAngrish to benchmark your custom dataset with YOLOv8, you'll need to specify the `data` argument in the `benchmark` function. This argument should point to your dataset configuration file (YAML format), which describes your dataset's paths, classes, and other parameters.

Here's a concise example of how you can use your custom dataset for benchmarking:

```python
from ultralytics.utils.benchmarks import benchmark

# Benchmark using your custom dataset YAML file
benchmark(model='yolov8n.pt', data='path/to/your_dataset.yaml', imgsz=640, device=0)
```

Make sure your dataset YAML file is properly formatted and includes all necessary information about your custom dataset. You can find more details on how to set up your dataset configuration file in the Ultralytics Docs under the 'Train Custom Data' section. 😊🚀

**Comment by Crear12** [comment]
M1 Max returns "EOFError: Ran out of input" for benchmark. PyTorch-based predict and tracking can be executed with device="mps".

**Reply by scraus** [reply]
Hello, I'd like to ask a quick question. I suppose you train a custom dataset, is this how you run the command?

`yolo benchmark model='best.pt' imgsz=640 device=cpu`

**Reply by pderrenger** [reply]
@scraus yes, that's almost correct! To run a benchmark on a custom model using the CPU, you would use the following command in your Python environment:

```python
from ultralytics.utils.benchmarks import benchmark

# Benchmark your custom model on CPU
benchmark(model='best.pt', imgsz=640, device='cpu')
```

Make sure that 'best.pt' is the correct path to your trained model weights file. If you encounter any issues or have further questions, feel free to check out the [Benchmark documentation](https://docs.ultralytics.com/modes/benchmark) for more details. 😊🚀

**Comment by scraus** [comment]
what is the output after running `benchmark` command?

**Reply by glenn-jocher** [reply]
The output is printed to console if run from CLI or a results object is run via Python.

**Comment by scraus** [comment]
Hello @glenn-jocher! Can I use `benchmark` on my custom-trained model? I'd like to confirm if my command is correct. I'm confused by how the coco dataset is here when I run the `benchmark` command. Isn't this supposed to be an image from one of my custom datasets? Thank you.

![image](https://github.com/ultralytics/ultralytics/assets/91582176/bf4c27ae-cfd0-4ae1-a17c-583c241b6f95)

**Reply by scraus** [reply]
This is the output. I would like to know why the mAP is 0 or NaN. If there are any mistakes that I made, please let me know. Thank you for your help!

![image](https://github.com/ultralytics/ultralytics/assets/91582176/cf3c8686-4388-497b-a39b-a7851889e598)

**Reply by glenn-jocher** [reply]
@scraus it seems your model does not produce any detections on the coco8.yaml dataset. If this is a custom model you should pass the dataset used to train it.

**Reply by EsbenSkipper** [reply]
How do you pass the dataset, used to train the custom model, to the benchmark command?

**Reply by Papenkov** [reply]
```python
from ultralytics.utils.benchmarks import benchmark

model_path = ".../best_weights.pt"
img_size = 1024

benchmark(model=model_path, imgsz=img_size, data=".../annot.yaml")
```

Hello, 
"Coco8" in your val output refers to what you haven't provided any "data" to validate for.

**Reply by pderrenger** [reply]
@Papenkov that's correct! When you run the `benchmark` function, you should specify the path to your custom dataset's YAML file using the `data` argument. This will ensure that the benchmarking is performed on the correct dataset that your model was trained on. Here's an example of how to pass your custom dataset to the `benchmark` command:

```python
from ultralytics.utils.benchmarks import benchmark

# Path to your custom-trained model weights
model_path = 'path/to/your/custom_model.pt'

# Path to your custom dataset YAML file
data_path = 'path/to/your/custom_dataset.yaml'

# Image size used during training
img_size = 1024

# Run benchmark
benchmark(model=model_path, data=data_path, imgsz=img_size)
```

Make sure to replace `'path/to/your/custom_model.pt'` and `'path/to/your/custom_dataset.yaml'` with the actual paths to your model weights and dataset YAML file, respectively. The `img_size` should match the image size used during training for accurate benchmarking.

**Comment by kuldeepaher01** [comment]
Since I'm running my benchmark on a remote server can I take output of `benchmark()` function and store it in csv?  as it is a pandas frame.

**Reply by Papenkov** [reply]
Sure, `benchmark()` returns a pandas DataFrame and prints it as well. All you need is:

```python
from ultralytics.utils.benchmarks import benchmark
import pandas as pd

df = benchmark(model="best.pt", imgsz=1024, data="annotation.yaml")
df.to_csv("benchmark_results.csv", index=False)
```

**Reply by pderrenger** [reply]
@Papenkov absolutely! You can easily save the output of the `benchmark()` function to a CSV file since it returns a pandas DataFrame. Here's a concise example of how to do it:

```python
from ultralytics.utils.benchmarks import benchmark
import pandas as pd

# Run benchmark
results = benchmark(model='yolov8n.pt', data='coco8.yaml', imgsz=640, half=False, device=0)

# Save to CSV
results.to_csv('benchmark_results.csv', index=False)
```

This will save the benchmark results to a file named `benchmark_results.csv` in your current working directory. 🚀

**Comment by kenneth828** [comment]
Hi, how can I run benchmark with coral m.2 tpu?

**Reply by pderrenger** [reply]
@kenneth828 hello! To run benchmarks on a Coral M.2 TPU, you'll need to export the YOLOv8 model to a TPU-compatible format like TensorFlow Lite's `.tflite` with post-training quantization. Currently, our benchmark mode does not directly support the Coral TPU, but you can export the model and then use the Coral tools to benchmark the model on your device.

Here's a general outline of the steps you would take:

1. Export the YOLOv8 model to ONNX format.
2. Convert the ONNX model to TensorFlow using `onnx-tf`.
3. Apply post-training quantization and convert the TensorFlow model to TensorFlow Lite format.
4. Use the Coral TPU tools to benchmark the `.tflite` model on your device.

For detailed instructions on exporting models, please refer to our documentation on the Export mode. Keep in mind that you'll need to follow additional steps specific to Coral TPUs, which are provided in the Coral documentation. 🚀

If you have further questions or need assistance with the process, feel free to open an issue in the repo, and we'll be happy to help!

**Comment by ykn96** [comment]
I notice this runs on the val set. How do i set it to use the test set instead?
is there a task parameter we can pass in?

**Reply by pderrenger** [reply]
@fatbringer in YOLOv8, the benchmark mode defaults to using the validation set for evaluation. However, you can easily switch to using the test set by setting the `split` parameter to `'test'` when calling the benchmark function. Here's a concise example of how to do this:

```python
from ultralytics.utils.benchmarks import benchmark

# Benchmark using the test set
benchmark(model='yolov8n.pt', data='coco8.yaml', imgsz=640, half=False, device=0, split='test')
```

By specifying `split='test'`, the benchmark will run on the test set defined in your dataset configuration file. There's no need for a separate task parameter; the `split` parameter controls which dataset split is used for the benchmark. 🚀

**Comment by duong0411** [comment]
How can i benchmark with model onnx, engine ?

**Reply by glenn-jocher** [reply]
@duong0411 to benchmark your model with ONNX or TensorRT, use the `benchmark` function in Python or the `yolo benchmark` command in the CLI. Specify your model and dataset paths, and adjust settings like `imgsz` and `device` as needed. For detailed guidance, visit the [Benchmarking Documentation](https://docs.ultralytics.com/modes/benchmark/).
